engine/consensus/Sealing_Readme.md

```diff
@@ -0,0 +1,109 @@
+- if receipts are received before the block, we will drop the receipt
+- If receipts are received after the block, but the previous result is missing, we will cache the receipt in the mempool
+- If previous result exists, we can validate the receipt, and store it in the execution tree mempool and storage.
+- After processing a receipt, we try to find child receipts from the cache, and process them.
+- When constructing a new block, the block builder reads receipts and results from the execution tree and add them to the new block’s payload
+- Processing finalized block is to prune pending receipts and execution tree mempool.
+- Matching core also contains logic for fetching missing receipts (method `requestPendingReceipts`)
+
+  Caution: this logic is not yet fully BFT compliant. The logic only requests receipts for blocks, where it has less than 2 consistent receipts. This is a temporary simplification and incompatible with the mature BFT protocol! There might be multiple consistent receipts that commit to a wrong result. To guarantee sealing liveness, we need to fetch receipts from *all*  ENs, whose receipts we don't have yet. There might be only a single honest EN and its result might occasionally get lost during transmission.
+
+- **Crash recovery:**
+    - Crash recovery is implemented in the consensus block builder (`module/builder/consensus/builder.go` method `repopulateExecutionTree`).
+    - During normal operations, we query the tree for "all receipts, whose results are derived from the latest sealed and finalized result". This requires the execution tree to know what the latest sealed and finalized result is.
+
+      The builder adds the result for the latest block that is both sealed and finalized, without any Execution Receipts. This is sufficient to create a vertex in the tree. Thereby, we can traverse the tree, starting from the sealed and finalized result, to find derived results and their respective receipts.
+
+    - In a second step, the block builder repopulates the builder adds all known receipts for any finalized but unsealed block to the execution tree.
+    - Lastly, the builder adds receipts from any valid block that descends from the latest finalized block.
+    - Note:
+
+      It is *strictly required* to add the result for the latest block that is both sealed and finalized. This is necessary because the ID of the latest result is given as a query to find all descending results during block construction. Without the sealed result, the block builder could not proceed.
+
+      In contrast, conceptually it is *not strictly required* for the builder to also add the receipts from the known blocks (finalized unsealed blocks and all pending blocks descending from the latest finalized block). This is because this matching engine *would* re-request these (method `requestPendingReceipts`) by itself. Nevertheless, relying on re-feting already known receipts would add a large latency overhead after recovering from a crash.
```

Comment:
```suggestion
      In contrast, conceptually it is *not strictly required* for the builder to also add the receipts from the known blocks (finalized unsealed blocks and all pending blocks descending from the latest finalized block). This is because this matching engine *would* re-request these (method `requestPendingReceipts`) by itself. Nevertheless, relying on re-fetching already known receipts would add a large latency overhead after recovering from a crash.
```